sympy/integrals/integrals.py

```diff
@@ -1596,3 +1603,61 @@ def line_integrate(field, curve, vars):
 
     integral = Integral(Ft, curve.limits).doit(deep=False)
     return integral
+
+
+def pole_reintegrate(f, x, antideriv,
+                     reintegrator=lambda f, x: integrate(f, x, conds="none")):
+    """
+    Takes an integrand f with integration variable x with an initially
+    computed antiderivative and checks for poles in the denominator. For each
+    of these poles, the integral is reevaluated, and the final integration
+    result is given in terms of a Piecewise.
+
+    See Also
+    ========
+
+    heurisch_wrapper
+    """
+    from sympy.solvers.solvers import solve, denoms
+
+    # We consider each denominator in the expression, and try to find
+    # cases where one or more symbolic denominator might be zero. The
+    # conditions for these cases are stored in the list slns.
+    slns = []
+    for d in denoms(antideriv):
+        try:
+            slns += solve(d, dict=True, exclude=(x,))
+        except NotImplementedError:
+            pass
+    if not slns:
+        return antideriv
+    slns = list(uniq(slns))
+    # Remove the solutions corresponding to poles in the original expression.
+    slns0 = []
+    for d in denoms(f):
+        try:
+            slns0 += solve(d, dict=True, exclude=(x,))
```

Comment:
Why do we have to solve here too? Can't we filter out the common denominators from `f` without calling `solve`? That would save a lot of unnecessary work if `f` has a lot of complicated symbolic constant expressions in denominators, which are unrelated to the integration variable. It's also possible for such expressions to not be solvable, in which case they would be filtered if I am reading this code correctly.

Reply:
This comes from #2112 but we can just check if `d.subs(already_computed_solution) == 0` but still looking at optimizations. Have a few changes coming regarding the comments.